adapters/repos/db/vector/hnsw/visited/list_set.go

```diff
@@ -0,0 +1,91 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2022 SeMI Technologies B.V. All rights reserved.
+//
+//  CONTACT: hello@semi.technology
+//
+
+package visited
+
+// ListSet is a reusable list with very efficient resets. Inspired by the C++
+// implementation in hnswlib it can be reset with zero memrory writes in the
+// array by moving the match target instead of altering the list. Only after a
+// version overflow do we need to actually reset
+//
+// The new implemtation uses a slice where the first element is reserved for the marker.
+// This allow us to use ListSet as a value (i.e. no pointer is required)
+// The marker (i.e. set[0]) allows for reusing the same list without having to zero all elements on each list reset.
+// Resetting the list takes place once the marker (i.e. set[0]) overflows
+type ListSet struct {
+	set []uint8 // set[0] is reserved for the marker (version)
+}
+
+//  Len returns the number of elements in the list.
+func (l ListSet) Len() int { return len(l.set) - 1 }
```

Comment:
If `len(l.set)` is `0`, would this return `-1`? or is there a mechanism in place elsewhere to prevent that from happening?

Reply:
The list is meant to be created using the constructor NewList() only. In this case the return value is always >= 0.